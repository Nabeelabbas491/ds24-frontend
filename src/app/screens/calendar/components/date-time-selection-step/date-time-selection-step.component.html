<div class="flex flex-row justify-start bg-neutral-n50">
  <div class="flex flex-col justify-start items-center">
    <ds-booking-user-profile [bookingTemplateSummary]="bookingTemplateSummary$ | async"></ds-booking-user-profile>
    <h4 class="font-medium my-3">{{ 'CALENDAR.SELECT_A_DATE' | translate }}</h4>
    <ds-calendar></ds-calendar>
    <div class="bg-neutral-n0 p-5">
      <ds-time-selection></ds-time-selection>
      <div class="mt-5" *ngIf="isDateTimeSelectionStepComplete$ | async">
        <button
          class="w-full h-10 py-2 px-4 bg-primary-p300-main text-neutral-n0 shadow-shadow-z200 text-center"
          (click)="nextStep()"
        >
          {{ 'BOOKING.BOOKING_OTHER.NEXT' | translate }}
        </button>
      </div>
    </div>
  </div>
</div>
