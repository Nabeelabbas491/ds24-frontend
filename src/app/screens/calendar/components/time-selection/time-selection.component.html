<h4 class="font-medium">{{ 'BOOKING.SELECT_TIME' | translate }}</h4>
<div class="mt-5" *ngIf="selectedDay$ | async as selectedDay">
  <form [formGroup]="form">
    <ds-select formControlName="timeZone" class="text-left">
      <ds-select-filter></ds-select-filter>
      <ds-form-label>{{ 'BOOKING.SELECT_TIME_ZONE' | translate }}</ds-form-label>
      <ds-select-option *ngFor="let timezone of timeZones$ | async" [value]="timezone.name">{{
        'TIMEZONES.' + timezone.name | translate
      }}</ds-select-option>
    </ds-select>
  </form>
  <h4 class="text-neutral-n500 mt-3 text-base">
    {{ selectedDay?.dayMonthYearTitle | dateFormat: 'full_without_year' }}
  </h4>
  <ng-template #timeslotslist>
    <div class="flex flex-row justify-start flex-wrap gap-3 mt-3" *ngIf="bookingTimeSlots$ | async as timeSlots">
      <p *ngIf="!timeSlots || timeSlots.length === 0; else timeslots" class="text-neutral-n300">
        {{ 'BOOKING.BOOKING_OTHER.NOT_AVAILABLE_AT_THIS_DATE' | translate }}
      </p>
      <ng-template #timeslots>
        <ds-chip
          [selectable]="true"
          [size]="'default'"
          [chipColor]="chipColor.LightPrimaryBlue"
          [selected]="(selectedTimeSlot$ | async) === item"
          (chipChange)="selectTimeSlot(item)"
          *ngFor="let item of timeSlots"
        >
          <ds-chip-text class="w-20 text-center">{{ item.startTime }}</ds-chip-text>
        </ds-chip>
      </ng-template>
    </div>
  </ng-template>
  <ds-spinner
    [color]="spinnerColor"
    class="mx-auto pl-6"
    *ngIf="isBookingTimeSlotsPending$ | async; else timeslotslist"
    [size]="50"
  ></ds-spinner>
</div>
