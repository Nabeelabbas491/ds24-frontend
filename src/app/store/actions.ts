import { actions as calendar } from './slices/calendar.store';
import { actions as booking } from './slices/booking.store';
import { actions as bookingAPI } from './slices/booking-api.store';
import { actions as vendor } from './slices/vendor.store';
import { actions as notification } from './slices/notification.store';
import { actions as viewBreakPoint } from './slices/view-breakpoint.store';
import { actions as vendorBookingTemplate } from './slices/vendor-booking-template.store';
import { actions as vendorSettings } from './slices/vendor-settings.store';
import { actions as vendorAppointment } from './slices/vendor-appointment.store';
import { actions as vendorCalendar } from './slices/vendor-calendar.store';
import { actions as vendorSchedule } from './slices/vendor-schedule.store';
import { actions as timeZoneAPI } from './slices/timezone-api.store';
import { actions as googleIntegration } from './slices/google-integration.store';

export {
  calendar,
  booking,
  bookingAPI,
  vendor,
  notification,
  viewBreakPoint,
  vendorBookingTemplate,
  vendorSettings,
  vendorAppointment,
  vendorCalendar,
  vendorSchedule,
  timeZoneAPI,
  googleIntegration,
};
