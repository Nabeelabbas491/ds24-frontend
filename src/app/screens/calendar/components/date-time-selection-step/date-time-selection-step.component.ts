import { Component } from '@angular/core';
import { Store } from '@ngrx/store';
import { selectors, actions } from './../../../../store';

@Component({
  selector: 'ds-date-time-selection-step',
  templateUrl: './date-time-selection-step.component.html',
})
export class DateTimeSelectionStepComponent {
  selectedDay$ = this.store.select(selectors.booking.selectSelectedDay);
  bookingTemplateSummary$ = this.store.select(selectors.booking.selectBookingTemplateSummary);
  isDateTimeSelectionStepComplete$ = this.store.select(selectors.booking.isDateTimeSelectionStepComplete);

  constructor(private store: Store) {}

  nextStep() {
    this.store.dispatch(actions.booking.selectStep({ step: 'booking_type' }));
  }
}
