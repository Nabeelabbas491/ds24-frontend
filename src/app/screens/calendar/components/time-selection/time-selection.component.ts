import { Component } from '@angular/core';
import { ChipColor, IconColor, IconName } from '@ds24/elements';
import { Store } from '@ngrx/store';
import { ReactiveLifecycles } from '@ds24/utilities';
import { FormBuilder, FormControl, FormGroup } from '@angular/forms';
import { takeUntil } from 'rxjs';

import * as actions from '../../../../store/actions';
import { selectors } from '../../../../store';
import { Slot } from './../../../../types/calendar-day.type';
@Component({
  selector: 'ds-time-selection',
  templateUrl: './time-selection.component.html',
})
export class TimeSelectionComponent extends ReactiveLifecycles {
  icons: typeof IconName = IconName;
  chipColor: typeof ChipColor = ChipColor;
  spinnerColor = IconColor.Primary300;

  selectedTimeZone$ = this.store.select(selectors.booking.selectTimeZone);
  selectedDay$ = this.store.select(selectors.booking.selectSelectedDay);
  selectedTimeSlot$ = this.store.select(selectors.booking.selectTimeSlot);

  isBookingTimeSlotsPending$ = this.store.select(selectors.bookingAPI.selectBookingTimeSlotsPending);
  bookingTimeSlots$ = this.store.select(selectors.bookingAPI.selectBookingTimeSlots);

  timeZones$ = this.store.select(selectors.timeZoneAPI.selectTimeZones);

  form: FormGroup = this.fb.group({
    timeZone: new FormControl(''),
  });

  constructor(
    private store: Store,
    private fb: FormBuilder,
  ) {
    super();
  }

  ngOnInit() {
    super.ngOnInit();

    this.selectedTimeZone$.pipe(takeUntil(this.ngOnDestroy$)).subscribe(x => {
      this.formControls.timeZone.setValue(x);
    });

    this.formControls.timeZone.valueChanges.pipe(takeUntil(this.ngOnDestroy$)).subscribe((timeZone: string) => {
      this.store.dispatch(actions.booking.timeZone({ timeZone }));
    });
  }
  selectTimeSlot(timeSlot: Slot) {
    this.store.dispatch(actions.booking.timeSlot({ timeSlot }));
  }

  get formControls(): any {
    return this.form.controls;
  }
}
