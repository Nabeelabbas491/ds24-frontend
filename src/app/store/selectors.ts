import { selectors as router } from './slices/router.store';
import { selectors as calendar } from './slices/calendar.store';
import { selectors as booking } from './slices/booking.store';
import { selectors as bookingAPI } from './slices/booking-api.store';
import { selectors as vendor } from './slices/vendor.store';
import { selectors as notification } from './slices/notification.store';
import { selectors as viewBreakPoint } from './slices/view-breakpoint.store';
import { selectors as vendorBookingTemplate } from './slices/vendor-booking-template.store';
import { selectors as vendorSettings } from './slices/vendor-settings.store';
import { selectors as vendorAppointment } from './slices/vendor-appointment.store';
import { selectors as vendorCalendar } from './slices/vendor-calendar.store';
import { selectors as vendorSchedule } from './slices/vendor-schedule.store';
import { selectors as timeZoneAPI } from './slices/timezone-api.store';
import { selectors as googleIntegration } from './slices/google-integration.store';

export {
  calendar,
  booking,
  bookingAPI,
  router,
  vendor,
  notification,
  viewBreakPoint,
  vendorBookingTemplate,
  vendorSettings,
  vendorAppointment,
  vendorCalendar,
  vendorSchedule,
  timeZoneAPI,
  googleIntegration,
};
