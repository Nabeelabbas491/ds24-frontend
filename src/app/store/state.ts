import { RouterReducerState } from '@ngrx/router-store';

import * as vendor from './slices/vendor.store';
import * as calendar from './slices/calendar.store';
import * as booking from './slices/booking.store';
import * as bookingAPI from './slices/booking-api.store';
import * as notification from './slices/notification.store';
import * as viewBreakPoint from './slices/view-breakpoint.store';
import * as vendorBookingTemplate from './slices/vendor-booking-template.store';
import * as vendorSettings from './slices/vendor-settings.store';
import * as vendorAppointment from './slices/vendor-appointment.store';
import * as vendorCalendar from './slices/vendor-calendar.store';
import * as vendorSchedule from './slices/vendor-schedule.store';
import * as timeZoneAPI from './slices/timezone-api.store';
import * as googleIntegration from './slices/google-integration.store';

/**
 * As mentioned, we treat each reducer like a table in a database. This means
 * our top level state interface is just a map of keys to inner state types.
 */
export interface State {
  [vendor.featureKey]: vendor.State;
  [notification.featureKey]: notification.State;
  [calendar.featureKey]: calendar.State;
  [booking.featureKey]: booking.State;
  [bookingAPI.featureKey]: bookingAPI.State;
  [viewBreakPoint.featureKey]: viewBreakPoint.State;
  [vendorBookingTemplate.featureKey]: vendorBookingTemplate.State;
  [vendorSettings.featureKey]: vendorSettings.State;
  [vendorAppointment.featureKey]: vendorAppointment.State;
  [vendorCalendar.featureKey]: vendorCalendar.State;
  [vendorSchedule.featureKey]: vendorSchedule.State;
  [timeZoneAPI.featureKey]: timeZoneAPI.State;
  [googleIntegration.featureKey]: googleIntegration.State;
  router: RouterReducerState;
}
