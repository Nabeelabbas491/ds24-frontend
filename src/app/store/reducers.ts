import { isDevMode } from '@angular/core';
import { MetaReducer, ActionReducerMap } from '@ngrx/store';
import { routerReducer } from '@ngrx/router-store';

import { State } from './state';

/**
 * Every reducer module's default export is the reducer function itself. In
 * addition, each module should export a type or interface that describes
 * the state of the reducer plus any selector functions. The `* as`
 * notation packages up all of the exports into a single object.
 */

import * as calendar from './slices/calendar.store';
import * as booking from './slices/booking.store';
import * as bookingAPI from './slices/booking-api.store';
import * as vendor from './slices/vendor.store';
import * as notification from './slices/notification.store';
import * as viewBreakPoint from './slices/view-breakpoint.store';
import * as vendorBookingTemplate from './slices/vendor-booking-template.store';
import * as vendorSettings from './slices/vendor-settings.store';
import * as vendorAppointment from './slices/vendor-appointment.store';
import * as vendorCalendar from './slices/vendor-calendar.store';
import * as vendorSchedule from './slices/vendor-schedule.store';
import * as timeZoneAPI from './slices/timezone-api.store';
import * as googleIntegration from './slices/google-integration.store';

/**
 * Our state is composed of a map of action reducer functions.
 * These reducer functions are called with each dispatched action
 * and the current or initial state and return a new immutable state.
 */
export const rootReducers: ActionReducerMap<State> = {
  [vendor.featureKey]: vendor.reducer,
  [calendar.featureKey]: calendar.reducer,
  [booking.featureKey]: booking.reducer,
  [bookingAPI.featureKey]: bookingAPI.reducer,
  [notification.featureKey]: notification.reducer,
  [viewBreakPoint.featureKey]: viewBreakPoint.reducer,
  [vendorBookingTemplate.featureKey]: vendorBookingTemplate.reducer,
  [vendorSettings.featureKey]: vendorSettings.reducer,
  [vendorAppointment.featureKey]: vendorAppointment.reducer,
  [vendorCalendar.featureKey]: vendorCalendar.reducer,
  [vendorSchedule.featureKey]: vendorSchedule.reducer,
  [timeZoneAPI.featureKey]: timeZoneAPI.reducer,
  [googleIntegration.featureKey]: googleIntegration.reducer,
  router: routerReducer,
};

/**
 * By default, @ngrx/store uses combineReducers with the reducer map to compose
 * the root meta-reducer. To add more meta-reducers, provide an array of meta-reducers
 * that will be composed to form the root meta-reducer.
 */
export const metaReducers: MetaReducer<State>[] = isDevMode() ? [] : [];
