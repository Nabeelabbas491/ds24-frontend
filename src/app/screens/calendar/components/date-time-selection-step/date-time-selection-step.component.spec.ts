import { ComponentFixture, TestBed } from '@angular/core/testing';

import { TranslateModule } from '@ngx-translate/core';
import { IconModule } from '@ds24/elements';
import { TimeSelectionComponent } from '../time-selection/time-selection.component';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { State, actions, selectors } from '../../../../store';
import { MemoizedSelector } from '@ngrx/store';
import { DateTimeSelectionStepComponent } from './date-time-selection-step.component';

describe('DateTimeSelectionStep', () => {
  let fixture: ComponentFixture<DateTimeSelectionStepComponent>;
  let instance: DateTimeSelectionStepComponent;
  let mockStore: MockStore<State>;

  let isDateTimeSelectionStepComplete: MemoizedSelector<State, boolean>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [DateTimeSelectionStepComponent, TimeSelectionComponent],
      imports: [TranslateModule.forRoot(), IconModule, FormsModule, ReactiveFormsModule],
      providers: [provideMockStore()],
    });
    fixture = TestBed.createComponent(DateTimeSelectionStepComponent);
    instance = fixture.componentInstance;
    mockStore = TestBed.inject(MockStore);

    isDateTimeSelectionStepComplete = mockStore.overrideSelector(
      selectors.booking.isDateTimeSelectionStepComplete,
      false,
    );

    fixture.detectChanges();

    jest.spyOn(mockStore, 'dispatch');
  });

  it('should compile', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('isBookingSelectionSepComplete should be set to true', () => {
    isDateTimeSelectionStepComplete.setResult(true);
    mockStore.refreshState();
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch next step action', () => {
    instance.nextStep();

    const action = actions.booking.selectStep({ step: 'booking_type' });

    expect(mockStore.dispatch).toHaveBeenLastCalledWith(action);
  });

  it('should create', () => {
    expect(instance).toBeTruthy();
  });
});
