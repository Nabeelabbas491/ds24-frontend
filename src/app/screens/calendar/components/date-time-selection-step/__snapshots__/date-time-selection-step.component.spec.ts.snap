// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimeSelectionStep isBookingSelectionSepComplete should be set to true 1`] = `
<ds-date-time-selection-step
  bookingTemplateSummary$={[Function _Store]}
  isDateTimeSelectionStepComplete$={[Function _Store]}
  selectedDay$={[Function _Store]}
  store={[Function _MockStore]}
>
  <div
    class="flex flex-row justify-start bg-neutral-n50"
  >
    <div
      class="flex flex-col justify-start items-center"
    >
      <ds-booking-user-profile />
      <h4
        class="font-medium my-3"
      >
        CALENDAR.SELECT_A_DATE
      </h4>
      <ds-calendar />
      <div
        class="bg-neutral-n0 p-5"
      >
        <ds-time-selection>
          <h4
            class="font-medium"
          >
            BOOKING.SELECT_TIME
          </h4>
        </ds-time-selection>
        <div
          class="mt-5"
        >
          <button
            class="w-full h-10 py-2 px-4 bg-primary-p300-main text-neutral-n0 shadow-shadow-z200 text-center"
          >
             BOOKING.BOOKING_OTHER.NEXT 
          </button>
        </div>
      </div>
    </div>
  </div>
</ds-date-time-selection-step>
`;

exports[`DateTimeSelectionStep should compile 1`] = `
<ds-date-time-selection-step
  bookingTemplateSummary$={[Function _Store]}
  isDateTimeSelectionStepComplete$={[Function _Store]}
  selectedDay$={[Function _Store]}
  store={[Function _MockStore]}
>
  <div
    class="flex flex-row justify-start bg-neutral-n50"
  >
    <div
      class="flex flex-col justify-start items-center"
    >
      <ds-booking-user-profile />
      <h4
        class="font-medium my-3"
      >
        CALENDAR.SELECT_A_DATE
      </h4>
      <ds-calendar />
      <div
        class="bg-neutral-n0 p-5"
      >
        <ds-time-selection>
          <h4
            class="font-medium"
          >
            BOOKING.SELECT_TIME
          </h4>
        </ds-time-selection>
      </div>
    </div>
  </div>
</ds-date-time-selection-step>
`;
